index.html

```diff
@@ -1453,80 +1453,560 @@ <h4>Recommended Practice</h4>
+          Content-Type: application/json
+          false
+          </pre>
+          <p>
+            If the property can not be read successfully then the Web Thing
+            MUST send an HTTP response with an HTTP error code which describes the
+            reason for the failure. E.g.
+          </p>
+          <ul>
+            <li><code>400 Bad Request</code></li>
+            <li><code>401 Unauthorized</code></li>
+            <li><code>403 Forbidden</code></li>
+            <li><code>404 Not Found</code></li>
+            <li><code>500 Internal Server Error</code></li>
+          </ul>
+        </section>
+        <section id="writeproperty">
+          <h5><code>writeproperty</code></h5>
+          <p>
+            The URL of a <code>Property</code> resource to be used when writing
+            the value of a property MUST be obtained from a Canonical TD by
+            locating a
+            <a href="https://www.w3.org/TR/wot-thing-description/#form">
+            <code>Form</code></a> inside the corresponding
+            <a href="https://www.w3.org/TR/wot-thing-description/#propertyaffordance">
+            <code>PropertyAffordance</code></a>
+            for which the value of its <code>op</code> member is
+            <code>writeproperty</code> and the
+            <a href="https://tools.ietf.org/html/rfc3986#section-3.1">URI
+            scheme</a> [[RFC3986]] of the value of its <code>href</code> member
+            is <code>http</code> or <code>https</code>.
+          </p>
+          <p>In order to write the value of a property, a Consumer MUST send
+            an HTTP request to a Web Thing with:
+            <ul>
+              <li>Method set to <code>PUT</code></li>
+              <li>URL set to the URL of the <code>Property</code> resource</li>
+              <li><code>Accept</code> header set to <code>application/json
+                </code></li>
+              <li><code>Content-Type</code> header set to <code>application/json
+                </code></li>
+              <li>A body with a requested new value for the property serialized
+                in JSON</li>
+            </ul>
+          </p>
+          <pre class="example">
+          PUT /things/lamp/properties/on HTTP/1.1
+          Host: mythingserver.com
+          Content-Type: application/json
+          Accept: application/json
+          true
+          </pre>
+          <p>
+            If a Web Thing receives an HTTP request following the format
+            above and the Consumer has permission to write the corresponding
+            property, then upon successfully writing the value of the
+            property it MUST send an HTTP response with:
+            <ul>
+              <li>Status code set to <code>200</code></li>
+              <li><code>Content-Type</code> header set to <code>application/json
+                </code></li>
+              <li>A body with the new value of the property serialized in JSON
+                </li>
```

Comment:
Is it really best practice that PUT returns a value?
I would have said that a PUT response should *not* return any body value

Reply:
There's nothing in the HTTP specification that says a response to a [PUT](https://tools.ietf.org/html/rfc7231#section-4.3.4) request _shouldn't_ contain a body. It just says:

> If the target resource does have a current representation and that representation is successfully modified in accordance with the state of the enclosed representation, then the origin server MUST send either a 200 (OK) or a 204 (No Content) response to indicate successful completion of the request.

A 204 response obviously shouldn't contain a body, but a 200 OK can. It's pretty common to return a body in HTTP-based APIs in my experience, but it probably comes down to taste. See this [post](https://github.com/w3c/wot-profile/pull/77) on StackOverflow for some of the pros and cons.

Pro:
- Provides an extra level of validation that the operation was successful, by echoing back the new state of the property

Con:
- Requires extra bandwidth to echo back information that the Consumer probably already knows

Note that an empty body is not valid JSON, so if the body is going to be empty implementors also need to make sure that the Content-Type header is not set to `application/json`. The same would be true if we use DELETE requests anywhere in the protocol binding though (e.g. to cancel an ongoing action).

Since the body is not strictly necessary in this particular case, I'd be fine with changing the response to 204 if there's a consensus around that?

What about PATCH requests though? If in future we support PATCH for partial updates of properties, would that return a body?

Reply:
MindSphere is following the same approach, e.g., as shown [here](https://developer.mindsphere.io/apis/advanced-assetmanagement/api-assetmanagement-api-swagger-3-14-0.html). What is ECHONET or Oracle Cloud doing?

Reply:
As far as I know, Oracle also returns a body similar to what MindSphere does. I personally do not like these since these are subprotocols put into HTTP and do not really do things the REST way. 

Regarding the current proposal to include the written value in the response, this again creates some sort of a subprotocol, i.e. I need to know that the value I got is the value I have written to the resource so that I can do something with it (validation I guess). Also, there is no way to indicate this properly in a TD for properties. The TD below would not change if the Thing returns such a response for a `writeproperty` operation:

```js
{
    "@context": "http://www.w3.org/ns/td",
    "title": "MyLampThing",
    "properties": {
        "status": {
            "type": "string",
            "readOnly":false,
            "forms": [{"href": "https://mylamp.example.com/status"}]
        }
    },
}
```

To go further, this was already discussed in an issue (could not find the number now but @danielpeintner had opened it I think ) about specifying return values for `writeproperty` operation and in the call it was decided that this would not needed.

Reply:
I think this is the issue, https://github.com/w3c/wot-thing-description/issues/875

Reply:
@egekorkan OK, as I said I don't mind changing the response to a PUT request to being a 204 with no content, though it sounds like existing implementations do consistently return a body.

With regards to defining a sub-protocol over HTTP, that's [exactly what I expect](https://w3c.github.io/web-thing-protocol/requirements/#http-sub-protocol) the "Core Profile" to do, by defining enough defaults that a Consumer to take a very simple Thing Description and interpret it as a much more complex (canonical) Thing Description with a rich set of interactions.

I think this will become particularly apparent when we come to defining the protocol binding for actions, which ideally could take the full set of interactions you proposed in https://github.com/w3c/wot-thing-description/issues/302#issuecomment-802159857 (`invokeaction`, `queryaction`, `updateaction`, `cancelaction`) and express them in a very simple form in a Thing Description, which the Consumer would then apply a set of defaults to to arrive at a much more complex canonical Thing Description describing a rich set of interactions.

I think this is in the spirit of the current set of defaults defined in section [8.3.1](https://www.w3.org/TR/wot-thing-description/#http-binding-assertions) of the Thing Description specification.

Reply:
From today's TD call: We had a look into ECHONET and we saw that ECHONET also echo the payload when the PUT method is applied (please see the slide [here](https://github.com/w3c/wot/blob/main/PRESENTATIONS/2021-03-online-f2f/2021-03-15-ECHONET-Lite-WebAPI-ECHONET-Consortium.pdf), slide 20)

Reply:
Thanks for the update @sebastiankb, this does seem quite common.

For reference, in the last Architecture call the most compelling reason to change to a 204 (No Content) response seemed to be the case where a property value payload is very large, e.g. for image or video properties.